src/Server/Model/AbstractToken.php

```diff
@@ -39,105 +40,126 @@
     /**
      * @var Client
      */
-    protected $client;
+    private $client;
 
     /**
      * @var TokenOwnerInterface
      */
-    protected $owner;
+    private $owner;
 
     /**
-     * @var DateTime
+     * @var DateTimeImmutable
      */
-    protected $expiresAt;
+    private $expiresAt;
 
     /**
      * @var array
      */
-    protected $scopes = [];
+    private $scopes = [];
 
     /**
-     * Set the token (either access or refresh token)
-     *
-     * @param  string $token
-     * @return void
+     * AbstractToken constructor.
      */
-    public function setToken($token)
+    private function __construct()
     {
-        $this->token = (string) $token;
     }
 
     /**
-     * Get the token
-     *
-     * @return string
+     * @param int                          $ttl
+     * @param TokenOwnerInterface|null     $owner
+     * @param Client|null                  $client
+     * @param string|string[]|Scope[]|null $scopes
+     * @return AbstractToken
      */
-    public function getToken(): string
-    {
-        return $this->token;
+    protected static function createNew(
+        int $ttl,
+        TokenOwnerInterface $owner = null,
+        Client $client = null,
+        $scopes = null
+    ): AbstractToken {
+
+        if (isset($scopes) && $scopes instanceof Scope) {
+            $scopes = $scopes->getName();
+        }
+
+        if (is_string($scopes)) {
+            $scopes = explode(' ', $scopes);
+        }
+
+        if (is_array($scopes)) {
+            foreach ($scopes as &$scope) {
+                $scope = $scope instanceof Scope ? $scope->getName() : (string) $scope;
+            }
+        }
+
+        $class = get_called_class();
+        $token = new $class();
```

Comment:
I didn't try but did you try to do $token = new static(); ? I didn't try it should be worth the try :).